src/static/js/chapter.js

```diff
@@ -0,0 +1,154 @@
+//Data Save can be set to on, so let's check it
+function dataSaverEnabled() {
+  let dataSaver = false;
+  if ('connection' in navigator) {
+    dataSaver = navigator.connection.saveData;
+    if (dataSaver) {
+      console.log('DataSaver is enabled');
+      gtag('event', 'data-saver', { 'event_category': 'user', 'event_label': 'enabled', 'value': 1 });
+    } else {
+      gtag('event', 'data-saver', { 'event_category': 'user', 'event_label': 'not-enabled', 'value': 0 });
+    }
+  }
+
+  return dataSaver;
+}
+
+//iOS causes Google Sheets to create a 6000 by 3700 canvas, which annoyingly isn't supported by iOS!
+//Let's check if we have large Canvas support (annoyingly no API for this!)
+function highResolutionCanvasSupported() {
+
+  let largeCanvasSupported = false;
+
+  try {
+    // Set large sized canvas dimensions and draw test rectangle
+    const cvs = document ? document.createElement('canvas') : null;
+    const ctx = cvs && cvs.getContext ? cvs.getContext('2d') : null;
+    cvs.width = 6000;
+    cvs.height = 3700;
+    ctx.fillRect(5999, 3699, 1, 1);
+    largeCanvasSupported = Boolean(ctx.getImageData(5999, 3699, 1, 1).data[3]) == 1;
+  }
+  catch (e) {
+    largeCanvasSupported = false;
+  }
+
+  if (largeCanvasSupported) {
+    gtag('event', 'hi-res-canvas', { 'event_category': 'user', 'event_label': 'supported', 'value': 1 });
+  } else {
+    console.log('High resolution canvas images are not supported');
+    gtag('event', 'hi-res-canvas', { 'event_category': 'user', 'event_label': 'not-supported', 'value': 0 });
+  }
+
+  return largeCanvasSupported;
+
+};
+
+//We use Google Sheets for detailed visualisations
+//Check for support and switch out images if supported
+function upgradeInteractiveFigures() {
+
+  try {
+    if (highResolutionCanvasSupported() && !dataSaverEnabled()) {
+
+      let sheetsAllowed = true;
+
+      const sheets_url = 'https://docs.google.com/spreadsheets/';
+
+      //Google Sheets is sometimes blocked by companies/proxies so let's check for that
+      //Unfortunately no easy way to check this as cannot make non-CORS request to
+      //Another domain and can't check status response when CORS domain :-(
+      //Below seems to work for my corporate proxy anyway
+      fetch(window.sheets_url, { method: 'HEAD', mode: 'no-cors' })

```

Comment:
I think this is causing an error in the console:

`HEAD http://127.0.0.1:8080/en/2019/undefined net::ERR_ABORTED 404 (NOT FOUND)`

Why would it appear as `/undefined`? And is there any way to prevent this from causing console errors?

Reply:
Typo, but it doesn't matter anyway. 😔See below.